napari/components/grid.py

```diff
@@ -32,6 +32,13 @@ class GridCanvas(EventedModel):
     enabled: bool = False
     stride: int = 1
     shape: Tuple[int, int] = (-1, -1)
+    stride_range: Tuple[
```

Comment:
this is what I was trying to say in https://github.com/napari/napari/pull/3084#discussion_r678634727 when I said we should be mirroring this in `GridCanvas`.  This is also a pydantic model, so no need to put things like `not` `minimum`, `range`, etc... _on_ the model itself.  make them contraints on the respective fields.  What I was trying to say is that it's seems natural to me that these fields would be exactly like the settings:

```py
class GridCanvas(EventedModel):
    grid_stride: conint(ge=-50, le=50, ne=0)
```

etc...

Reply:
(and the regular pyandtic way to access those constraints):

```python
field = GridCanvas.__fields__['shape']
for sf in field.sub_fields:  # if it's a Tuple field, it will have subfields
    print(sf.type_.ge, sf.type_.ne)  # ge and ne are properties of the field type_

```